Feat: keychain rotate passphrase

zeim839 opened the pull request:
This is a rough implementation supporting [rotating keychain passphrases](https://github.com/libp2p/js-libp2p/issues/728). I need some suggestions on how to proceed. I haven't received a reply so I've copied the original conversation here for reference:

```js
async changeKeychainPassword(oldPassword, newPassword){
    if (typeof oldPassword !== 'string' || typeof newPassword !== 'string') {
      throw new Error(`Invalid pass type '${typeof oldPassword}'`);
    }
    if (newPassword.length < 20) {
      throw new Error('pass must be least 20 characters')
    }
    // TODO: Need to decide on how to import/generate opts
    const newDek = newPassword
      ? crypto.pbkdf2(
        newPassword,
        this.opts.dek.salt,
        this.opts.dek.iterationCount,
        this.opts.dek.keyLength,
        this.opts.dek.hash)
      : ''
    const oldDek = privates.get(this).dek
    privates.set(this, { "dek":newDek })
    var keys = await this.listKeys()
    await keys.forEach(async key =>{
      // TODO: Decide how to handle deleting and importing the "self" key. 
          // importKey and removeKey throw an error when handling "self"
      if (key.name != "self"){
        var keyAsPEM = await this._getPrivateKey(key.name)
        await this.removeKey(key.name)
        this.importKey(key.name, keyAsPEM, oldDek)
      }
    })
  }

```
[src/keychain/index.js](https://github.com/zeim839/js-libp2p/blob/feat/keychain-change-passphrase/src/keychain/index.js#L507)

What's left is to decide how to handle the "self" key (`this.importKey()` and `this.removeKey()` throw an error on "self") and how to handle changing passwords in the context of src/index.js (see [here](https://github.com/zeim839/js-libp2p/blob/feat/keychain-change-passphrase/src/index.js#L209)). I'm trying to be cautious/minimal with the changes I make so I'd appreciate some suggestions on how to proceed.

vasco-santos commented on src/index.js:
```diff
@@ -207,6 +207,7 @@ class Libp2p extends EventEmitter {
     }
 
     // Create keychain
+      // TODO: Decide how to change these options when the keychain password is being changed
```
Do you mean the `const keychainOpts = Keychain.generateOptions()`? I think we can just run it again to get the options. What do you think?

zeim839 replied:
Yes that sounds good to me.

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +503,36 @@ class Keychain {
     }
   }
+
+  async changeKeychainPassword(oldPassword, newPassword){
```
```suggestion
  async rotateKeychainPass (oldPass, newPass){
```

Perhaps rotate is a better naming, and we use pass everywhere, so I think this should also use `pass`

vasco-santos replied:
We probably need to specify the key we want to rotate?

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +503,36 @@ class Keychain {
       return throwDelayed(errcode(new Error(`Key '${name}' does not exist. ${err.message}`), 'ERR_KEY_NOT_FOUND'))
     }
   }
+
```
Please add `jsdoc` here

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +503,36 @@ class Keychain {
+    var keys = await this.listKeys()
+    await keys.forEach(async key =>{
+      // TODO: Decide how to handle deleting and importing the "self" key. 
+          // importKey and removeKey throw an error when handling "self"
```
yeah, we should not change this behaviour. I would say we should create internal methods to handle the common logic

vasco-santos reviewed (commented):
Thanks for the PR @zeim839 

I think I addressed all the comments

zeim839 commented:
@vasco-santos Please take a look at the new changes and let me know what you think. Regarding tests, I've incorporated testing for validation but I'm not exactly sure how to go about the following:
```js
it('can rotate keychain passphrase', async () => {  
    //TODO...
})
```

zeim839 commented:
I think I've come up with a possible solution. If its ok with you then everything should be good to go. I also decided to create a new keychain [here](https://github.com/libp2p/js-libp2p/pull/944/commits/825d70c1718f56700c20b69fa0418b158cf022c3#diff-86874834813c75bef940feea4eabc092bb90aeb6dd04173bbfa55ca8e7d766b1R497) (instead of using ks) to avoid influencing any other/future tests that dont use the new rotated password.

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +503,57 @@ class Keychain {
+   */
+  async rotateKeychainPass (oldPass, newPass){
+    if (typeof oldPass !== 'string' || typeof newPass !== 'string') {
+      throw new Error(`Invalid pass type '${typeof oldPass}'`);
```
Probably not essential, but what do you think about using `throwDelayed` like the rest of the keychain functions? And please add an `errCode`

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +503,57 @@ class Keychain {
+      throw new Error(`Invalid pass type '${typeof oldPass}'`);
+    }
+    if (newPass.length < 20) {
+      throw new Error('pass must be least 20 characters')
```
Same as above

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +503,57 @@ class Keychain {
+    if (newPass.length < 20) {
+      throw new Error('pass must be least 20 characters')
+    }
+    this.opts = {
```
Can we add a log before this with `recreating keychain` and after the keys iteration `keychain reconstructed` to be consistent with what we do on creation?

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +503,57 @@ class Keychain {
+        this.opts.dek.keyLength,
+        this.opts.dek.hash)
+      : ''
+    privates.set(this, { "dek":newDek })
```
```suggestion
    privates.set(this, { "dek": newDek })
```

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +503,57 @@ class Keychain {
+
+    var keys = await this.listKeys()
+    await keys.forEach(async key =>{
+      var keyAsPEM = await this._getPrivateKey(key.name)
```
```suggestion
      const keyAsPEM = await this._getPrivateKey(key.name)
```

vasco-santos replied:
this PEM was created with the old dek https://github.com/libp2p/js-libp2p/blob/v0.31.5/src/keychain/index.js#L238. You can look into the export key to see how to get the private key from pem using the previous `dek` and then you need to encrypt it once again as it is done in createKey

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +503,57 @@ class Keychain {
+      : ''
+    privates.set(this, { "dek":newDek })
+
+    var keys = await this.listKeys()
```
```suggestion
    const keys = await this.listKeys()
```

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +503,57 @@ class Keychain {
+
+
+      // Import key with new pass
+      var keyInfo = {
```
```suggestion
      const keyInfo = {
```

vasco-santos reviewed (changes requested):
I did an initial iteration on the PR. The new pm does not seem to be created with the new password. I left inline comments on how I think it would look like.

Also, before submitting the PR try to get the CI green please 🙏🏼  I added small inline suggestion to fix the lint, but I probably did not get to all of them. You should run both `npm run lint` and `npm run test`

Thanks @zeim839

zeim839 commented:
> Also, before submitting the PR try to get the CI green please 🙏🏼 I added small inline suggestion to fix the lint, but I probably did not get to all of them. You should run both `npm run lint` and `npm run test`

Sorry about that, I shouldve read the contrib guidelines more carefully. Anyway, I've ran lint and fixed any issues that I could trace back to me but I'm not sure If I can even get the CI to green because linting fails regardless of my edits. I've fixed the mistakes from the previous review and wrote the test as described. 

However, I keep getting `"Error: Cannot read the key, most likely the password is wrong or not a RSA key"` [here](https://github.com/libp2p/js-libp2p/pull/944/files#diff-6bbef5dd5eeccbdf00ec112882c711b59ef3db822873cb85f9244acf9a1e687dR542) when testing on browser, despite it working fine in test:node. Am I missing/overlooking something?

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +506,60 @@ class Keychain {
+    log('recreating keychain')
+    const oldDek = privates.get(this).dek
+    this.opts.pass = newPass
+    const newDek = newPass
```
```suggestion
    // eslint-disable-next-line constant condition
    const newDek = newPass
```

vasco-santos commented on test/keychain/keychain.spec.js:
```diff
@@ -492,6 +493,84 @@ describe('keychain', () => {
+            options.dek.hash)
+          : ''
+
+        const newDek = 'newInsecurePassphrase'
```
```suggestion
        // eslint-disable-next-line no-constant-condition
        const newDek = 'newInsecurePassphrase'
```

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +506,61 @@ class Keychain {
+    log('recreating keychain')
+    const oldDek = privates.get(this).dek
+    this.opts.pass = newPass
+    // eslint-disable-next-line constant condition
```
```suggestion
```

vasco-santos commented:
> Sorry about that, I shouldve read the contrib guidelines more carefully. Anyway, I've ran lint and fixed any issues that I could trace back to me but I'm not sure If I can even get the CI to green because linting fails regardless of my edits. I've fixed the mistakes from the previous review and wrote the test as described.

No problem, thanks for looking into it. I already submitted a suggestion fixing the lint. I will try to replicate the browser issue and get back to you

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +506,60 @@ class Keychain {
+      : ''
+    privates.set(this, { dek: newDek })
+    const keys = await this.listKeys()
+    await keys.forEach(async key => {
```
```suggestion
    for (const key of keys) {
```

this `await` does not work like this and the forEach does not return a promise. So this means `rotateKeychainPass` would return before running the cycle. As a side effect, in the browser tests were failing because it was slower + the usage of `setTimeout` in test

Note that the suggestion change below is also needed to close the for cycle correctly

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +506,60 @@ class Keychain {
+      batch.put(DsName(key.name), uint8ArrayFromString(keyAsPEM))
+      batch.put(DsInfoName(key.name), uint8ArrayFromString(JSON.stringify(keyInfo)))
+      await batch.commit()
+    })
```
```suggestion
    }
```

vasco-santos commented on test/keychain/keychain.spec.js:
```diff
@@ -492,6 +493,85 @@ describe('keychain', () => {
+      await kc.createKey('keyCreatedWithOldPassword', 'rsa', 2048)
+      await kc.rotateKeychainPass(oldPass, 'newInsecurePassphrase')
+
+      setTimeout(async () => {
```
With the above change, this `setTimeout` can be removed

vasco-santos commented on test/keychain/keychain.spec.js:
```diff
@@ -492,6 +493,85 @@ describe('keychain', () => {
+          : ''
+
+        // Dek with old password should not work:
+        expect(await kc.importKey('keyWhosePassChanged', pem, oldDek)).to.throw()
```
```suggestion
        await expect(kc.importKey('keyWhosePassChanged', pem, oldDek))
          .to.eventually.be.rejected()
```

vasco-santos commented on test/keychain/keychain.spec.js:
```diff
@@ -492,6 +493,85 @@ describe('keychain', () => {
+        // Dek with old password should not work:
+        expect(await kc.importKey('keyWhosePassChanged', pem, oldDek)).to.throw()
+        // Dek with new password should work:
+        expect(await kc.importKey('keyWhosePasswordChanged', pem, newDek)).to.have.property('name', 'keyWhosePasswordChanged')
```
```suggestion
        await expect(kc.importKey('keyWhosePasswordChanged', pem, newDek))
          .to.eventually.have.property('name', 'keyWhosePasswordChanged')
```

vasco-santos reviewed (changes requested):
Thanks for the new iteration. This is mostly done, I tried the changes I suggested below and I got everything to work.

Moreover, I added a lint skip for the error you had, as the linter was complaining for the ternary condition in the constant in the test, which is fine. When you add the suggestion, please check if the linting is still fine, given that the suggestion spaces could break the spacing.

Also, you will need to pull the changes done in the commit that fixed the linting
when you add the suggestions

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +506,60 @@ class Keychain {
+      const batch = this.store.batch()
+      batch.delete(DsName(key.name))
+      batch.delete(DsInfoName(key.name))
+      await batch.commit()
+
```
```suggestion
```

The last commit should be enough right?

zeim839 replied:
I thought so too at first but I believe there's some sort of issue where batch cannot both delete and create a key of the same name without separate commits. When I remove the first `batch.commit()` and run `rotateKeychainPass()`, the datastore appears to be completely empty like so: `MemoryDatastore { data: {} }`. If I try `batch.put()` with a different key name, the operation seems to work as expected (i.e the old key gets deleted and the new key is appended). 

Try running the below and let me know If I'm missing something or if this is a legitimate issue:
```js
await kc.createKey('keyCreatedWithOldPassword', 'rsa', 2048)
const dsnameData = new Key('/pkcs8/keyCreatedWithOldPassword')
const dsnameInfo = new Key('/info/keyCreatedWithOldPassword')
const batch = ds.batch()

// Delete key data and info
batch.delete(dsnameData)
batch.delete(dsnameInfo)

// Append key data and info
await batch.put(dsnameData, uint8ArrayFromString('pem'))
await batch.put(dsnameInfo, uint8ArrayFromString('info'))

await batch.commit()
console.log(ds) // Output: MemoryDatastore { data: {} }
```

vasco-santos replied:
Thanks for the details. I see, but rethinking this. A datastore will always be indexed by a key, so the `delete` is redundant as the `put` will write over it

vasco-santos replied:
Added two suggestions to get the delete out again.

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +506,60 @@ class Keychain {
+
+      // Remove key with old pass
+      const batch = this.store.batch()
+      batch.delete(DsName(key.name))
+      batch.delete(DsInfoName(key.name))
+      await batch.commit()
+
```
```suggestion
```

vasco-santos commented on src/keychain/index.js:
```diff
@@ -503,6 +506,60 @@ class Keychain {
+      const password = newDek.toString()
+      const keyAsPEM = await privateKey.export(password)
+
+      // Remove key with old pass
```
```suggestion
      // Update stored key
```

zeim839 commented:
@vasco-santos Thanks for the help, I wasn't aware of that. I've removed the redundancy and everything works as expected now.

vasco-santos reviewed (approved):
LGTM!
Thanks for landing this @zeim839

vasco-santos commented:
Released in `0.31.6`

zeim839 commented:
@vasco-santos Thanks for all the help!
